builtin/logical/database/backend.go

```diff
@@ -339,18 +386,45 @@ func (b *databaseBackend) CloseIfShutdown(db *dbPluginInstance, err error) {
 
 // clean closes all connections from all database types
 // and cancels any rotation queue loading operation.
+// It spawns goroutines to close the databases since these
+// are not guaranteed to finish quickly.
 func (b *databaseBackend) clean(ctx context.Context) {
-	// invalidateQueue acquires it's own lock on the backend, removes queue, and
+	cleanupCtx, cancel := context.WithDeadline(ctx, time.Now().Add(cleanupMaxWaitTime))
+	defer cancel()
+
+	// invalidateQueue acquires its own lock on the backend, removes queue, and
 	// terminates the background ticker
 	b.invalidateQueue()
 
 	b.Lock()
 	defer b.Unlock()
 
-	for _, db := range b.connections {
-		db.Close()
-	}
+	// copy all connections so we can close asynchronously
+	connectionsCopy := b.connections
 	b.connections = make(map[string]*dbPluginInstance)
+
+	// we will try to wait for all the connections to close
+	// ... but not too long
+	sem := semaphore.NewWeighted(int64(len(connectionsCopy)))
```

Comment:
This seems like a useful pattern that I have seen messed up a lot, and with [a lot](https://stackoverflow.com/questions/32840687/timeout-for-waitgroup-wait) of bad answers for on the internet; I'm tempted to move this out to a library since [I don't see one already made](https://pkg.go.dev/search?q=waitgroup) (what I want is something like a "TimedWaitGroup" or "WaitGroupWithTimeout".